### ISSUE TITLE
Support for S3 catalog to work with S3 Tables

### ISSUE THREAD

**Issue by nicor88** [apache/iceberg-python#1404]
### Feature Request / Improvement

Amazon S3 tables have being launched, see [this](https://aws.amazon.com/blogs/aws/new-amazon-s3-tables-storage-optimized-for-analytics-workloads/), and looks like that S3 tables have a managed iceberg catalog.

Based on https://github.com/awslabs/s3-tables-catalog looks like that they build an S3 catalog wrapper using java, but it will be relevant to being able to write to those S3 tables via pyiceberg.

Base on my understanding once an S3 table is created you can put any object in the S3 warehouse location of the table, and then using [update_table_metadata_location](https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/s3tables/client/update_table_metadata_location.html) to point the s3 table to the right location.

**Comment by kevinjqliu** [comment]
Thanks for raising this @nicor88! Would you be interested to contribute this feature?

**Comment by kevinjqliu** [comment]
The catalog implementation can be found here 
https://github.com/awslabs/s3-tables-catalog/blob/adfeece9873f06364a4a093bfedacb5efe4a952b/src/software/amazon/s3tables/iceberg/S3TablesCatalog.java

**Comment by nlm4145** [comment]
I also would be interested in this feature.

**Comment by nicor88** [comment]
@kevinjqliu Unfortunately I don't have the capacity at the moment to contribute to this feature.
I would nevertheless be available to look at the PR and test the implementation.

**Comment by petehanssens** [comment]
Super keen to see this happen too!

**Comment by felixscherz** [comment]
I created an intial PR #1429 where I am currently working on supporting table creation. I ran into the same issue that @nicor88 described and could work around it by setting `overwrite=True` for now.
However, now I get a different error during the write operation for the table metadata:
> AWS Error UNKNOWN (HTTP status 400) during CompleteMultipartUpload operation: Unable to parse ExceptionName: S3TablesUnsupportedHeader Message: S3 Tables does not support the following header: x-amz-api-version value: 2006-03-01

I'm currently going through the pyarrow S3FileSystem implementation to see where this header is being introduced.

**Comment by jamesbornholt** [comment]
@felixscherz thanks for catching this (and thanks to everyone who's interested in building S3 Tables support for PyIceberg!). We're working on an S3-side fix for the `x-amz-api-version` exception you're seeing; hoping to have that out soon.

**Comment by buremba** [comment]
@jamesbornholt Great to hear that you folks are keeping an eye on here! 
Sorry if this is not the right channel to ask the question but considering S3Tables is a mix of catalog + storage layer, is there any plan to provide Iceberg REST compatibility on top of S3Tables in addition to current API?

IMO that would help accelerate the adoption a lot, otherwise all the Iceberg implementations will need to integrate with S3Tables separately and I have a feeling that maintenance will be non-trivial.

**Comment by soumilshah1995** [comment]
+1

**Comment by felixscherz** [comment]
> @felixscherz thanks for catching this (and thanks to everyone who's interested in building S3 Tables support for PyIceberg!). We're working on an S3-side fix for the `x-amz-api-version` exception you're seeing; hoping to have that out soon.

I just re-ran the tests using PyArrowFileIO and it seems to be fixed now, thank you!

**Comment by soumilshah1995** [comment]
can we have full example code ?

**Comment by felixscherz** [comment]
@soumilshah1995 once #1429 is merged, an example would be:

```python
from pyiceberg.catalog.s3tables import S3TablesCatalog
import pyarrow as pa


table_bucket_arn: str = "..."
aws_region: str = "..."

properties = {"s3tables.warehouse": table_bucket_arn, "s3tables.region": aws_region}
catalog = S3TablesCatalog(name="s3tables_catalog", **properties)

database_name = "prod"

catalog.create_namespace(namespace=database_name)

pyarrow_table = pa.Table.from_arrays(
    [
        pa.array([None, "A", "B", "C"]),
        pa.array([1, 2, 3, 4]),
        pa.array([True, None, False, True]),
        pa.array([None, "A", "B", "C"]),
    ],
    schema=pa.schema(
        [
            pa.field("foo", pa.large_string(), nullable=True),
            pa.field("bar", pa.int32(), nullable=False),
            pa.field("baz", pa.bool_(), nullable=True),
            pa.field("large", pa.large_string(), nullable=True),
        ]
    ),
)

identifier = (database_name, "orders")
table = catalog.create_table(identifier=identifier, schema=pyarrow_table.schema)
table.append(pyarrow_table)

```

I'm currently working on implementing AWS S3 Tables for the moto library https://github.com/getmoto/moto/pull/8470 so once that is merged, we can improve the tests and merge #1429

**Comment by kevinjqliu** [comment]
I think that snippet would be great as part of the docs for S3 Tables :)

**Comment by felixscherz** [comment]
I will add that to the docs:) currently focusing on the moto side of things:)

**Comment by madsenwattiq** [comment]
Is there a timeline for getting this feature merged into a release?  I know you were working on getting the testing framework on the moto side completed, but that looks like it was merged awhile back....we're eagerly awaiting s3 table support!  Thank you!

**Comment by felixscherz** [comment]
Hi, the PR for this feature is done and in review at the moment: https://github.com/apache/iceberg-python/pull/1429 :)

**Comment by raphaelauv** [comment]
https://aws.amazon.com/about-aws/whats-new/2025/03/amazon-s3-tables-apache-iceberg-rest-catalog-apis/

aws s3 table compatible with HTTP rest catalog api , I did not try

**Comment by felixscherz** [comment]
I guess that makes #1429 obsolete:D
Supporting the REST Catalog seems like a logical choice for AWS:)

**Comment by kevinjqliu** [comment]
Thanks for all the great work you did here @felixscherz! it's a good reference for anyone looking to use the s3 tables specific apis.

Closing this since pyiceberg can access S3 Tables via the REST catalog client

**Comment by srstrickland** [comment]
Hi @kevinjqliu , can we consider reopening this issue considering this: https://github.com/apache/iceberg-python/pull/1429#issuecomment-3423334847

In summary, supporting the native s3table interface will insulate us from bugs like these in glue. Schema evolution has been effectively broken for over a month and we're still waiting for a fix to be deployed to glue rest API. There is no viable workaround today, other than moving away from pyiceberg. We must either wait for AWS to deploy a fix or for pyiceberg to support the native interface 

It was my understanding that the glue/rest interface was a stopgap and there was still a plan to support s3tables natively. Is that no longer the case?

Appreciate your thoughts on this 🙏

**Comment by kevinjqliu** [comment]
Hi @srstrickland thanks for raising this. 

Seems like this is an issue with the server side implementation of the iceberg rest catalog interface, and it looks like the glue team is aware based on https://github.com/apache/iceberg-python/issues/2511#issuecomment-3402084041

> We must either wait for AWS to deploy a fix or for pyiceberg to support the native interface

Do you know if a native s3tables implementation would resolve this issue? #1429 has the `commit_table` function implemented and we can test to see if it can unblock this particular issue. 

> It was my understanding that the glue/rest interface was a stopgap and there was still a plan to support s3tables natively. Is that no longer the case?

In my opinion, the rest interface is the preferred over specific implementations. This takes the maintenance burden away from pyiceberg. Client and server should speak the same [REST protocol](https://iceberg.apache.org/rest-catalog-spec/). The pyiceberg client wont need to worry about specific implementations. It can reuse the same logic for the glue rest catalog and every other rest catalog.
